### PR TITLE
Realm Tasks Migration to Swift 3

### PR DESCRIPTION
**Pull request by TimOliver**
A non-trivial PR; this converts the entire codebase of Realm Tasks for both macOS and iOS to Swift 3, allowing for the latest dependencies to subsequently be integrated.

There may still be some code issues in certain places, and certain things might be able to be converted more cleanly (There are a few `NSError` to `Error` compatibility sections that come to mind), but for now, I'd love a second pair of eyes to have a look at it so far. :)

/cc @jpsim @stel

### CODE REVIEW

File: RealmTasks Apple/RealmTasks iOS/TableViewCell.swift [hunk 1]
```diff
@@ -73,22 +73,22 @@ final class TableViewCell<Item: Object where Item: CellPresentable>: UITableView
     var presenter: CellPresenter<Item>!
 
     // Private Properties
-    private var originalDoneIconCenter = CGPoint()
-    private var originalDeleteIconCenter = CGPoint()
+    fileprivate var originalDoneIconCenter = CGPoint()
```
Comment (stel) [inline RealmTasks Apple/RealmTasks iOS/TableViewCell.swift]: These properties are not used in any extensions, so the could still be `private`

File: RealmTasks Apple/RealmTasks iOS/TablePresenter.swift [hunk 2]
```diff
@@ -67,154 +67,154 @@ class TablePresenter<Parent: Object where Parent: ListPresentable>: NSObject,
...
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
         return items.count
     }
 
-    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
-        let cell = viewController.tableView.dequeueReusableCellWithIdentifier("cell", forIndexPath: indexPath) as! TableViewCell<Parent.Item>
-
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = viewController.tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath) as! TableViewCell<Parent.Item>
+
```
Comment (stel) [inline RealmTasks Apple/RealmTasks iOS/TablePresenter.swift]: Whitespaces

File: RealmTasks Apple/RealmTasks macOS/ContainerViewController.swift [hunk 3]
```diff
@@ -34,29 +34,27 @@ class ContainerViewController: NSViewController {
...
-            visualEffectView.state = .Active
+            visualEffectView.state = .active
         }
     }
 
     @IBAction func showAllLists(sender: AnyObject?) {
-        let rootList = try! Realm().objects(TaskListList.self).first!
-
-        presentViewControllerForList(rootList)
+        let rootList: TaskListList = try! Realm().objects(TaskListList.self).first!
```
Comment (stel) [inline RealmTasks Apple/RealmTasks macOS/ContainerViewController.swift]: It's not necessary to specify the type here implicitly

File: RealmTasks Apple/RealmTasks macOS/ContainerViewController.swift [hunk 4]
```diff
@@ -34,29 +34,27 @@ class ContainerViewController: NSViewController {
...
-                visualEffectView.material = .Dark
+                visualEffectView.material = .dark
             }
 
-            visualEffectView.state = .Active
+            visualEffectView.state = .active
         }
     }
 
     @IBAction func showAllLists(sender: AnyObject?) {
```
Comment (stel) [inline RealmTasks Apple/RealmTasks macOS/ContainerViewController.swift]: `@IBAction`s should become `@IBAction func methodName(_ sender: AnyObject?)`

File: RealmTasks Apple/RealmTasks macOS/ItemCellView.swift [hunk 5]
```diff
@@ -115,7 +115,7 @@ class ItemCellView: NSTableCellView {
         setupUI()
         setupGestures()
 
-        setTrackingAreaWithRect(bounds, options: [.MouseEnteredAndExited, .ActiveInKeyWindow])
+        setTrackingAreaWithRect(rect: bounds, options: [.mouseEnteredAndExited, .activeInKeyWindow])
```
Comment (stel) [inline RealmTasks Apple/RealmTasks macOS/ItemCellView.swift]: Should probably be renamed to `setTrackingArea(with rect:, options:)` and `@discardableResult` should be added to avoid warnings.

File: RealmTasks Apple/RealmTasks macOS/ListViewController.swift [hunk 6]
```diff
@@ -160,26 +164,26 @@ final class ListViewController<ListType: ListPresentable where ListType: Object>
...
+        return validateSelector(selector: toolbarItem.action!)
     }
 
-    override func validateMenuItem(menuItem: NSMenuItem) -> Bool {
-        return validateSelector(menuItem.action)
+    override func validateMenuItem(_ menuItem: NSMenuItem) -> Bool {
+        return validateSelector(selector: menuItem.action!)
     }
 
     private func validateSelector(selector: Selector) -> Bool {
```
Comment (stel) [inline RealmTasks Apple/RealmTasks macOS/ListViewController.swift]: Should it be renamed to `validateSelector(_ selector: Selector)`?

File: RealmTasks Apple/Podfile [hunk 7]
```diff
@@ -4,8 +4,7 @@ abstract_target 'RealmTasks' do
     use_frameworks!
     
     pod 'RealmSwift', '~> 2.1.0'
-    
-    pod 'Cartography', '~> 0.7'
+    pod 'Cartography'
```
Comment (jpsim) [inline RealmTasks Apple/Podfile]: Why was the version specifier lost? Should be `pod 'Cartography', '~> 1.0'`

File: RealmTasks Apple/RealmTasks Shared/RealmColor.swift [hunk 8]
```diff
@@ -54,13 +54,13 @@ extension Color {
     }
 }
 
-extension CollectionType where Generator.Element == Color, Index == Int {
+extension Collection where Iterator.Element == Color, Index == Int {
     func gradientColorAtFraction(fraction: Double) -> Color {
         // Ensure offset is normalized to 1
-        let normalizedOffset = max(min(fraction, 1), 0)
+        let normalizedOffset = Swift.max(Swift.min(fraction, 1.0), 0.0)
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks Shared/RealmColor.swift]: why are these changes necessary?

File: RealmTasks Apple/RealmTasks Shared/RealmColor.swift [hunk 9]
```diff
@@ -54,13 +54,13 @@ extension Color {
...
 
         // Work out the 'size' that each color stop spans
-        let colorStopRange = 1 / Double(count - 1)
+        let colorStopRange = 1.0 / Double(self.count.toIntMax() - 1)
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks Shared/RealmColor.swift]: why are these changes necessary?

Comment (jpsim) [inline RealmTasks Apple/RealmTasks Shared/RealmColor.swift]: `toIntMax()` in particular looks very wrong.

File: RealmTasks Apple/RealmTasks Shared/SharedFunctionality.swift [hunk 10]
```diff
@@ -48,18 +48,18 @@ private func setDefaultRealmConfigurationWithUser(user: SyncUser) {
...
                 let listReferenceIDs = NSCountedSet(array: items.map { $0.id })
-                for id in listReferenceIDs where listReferenceIDs.countForObject(id) > 1 {
+                for id in listReferenceIDs where listReferenceIDs.count(for: id) > 1 {
                     let id = id as! String
-                    let indexesToRemove = items.enumerate().flatMap { index, element in
+                    let indexesToRemove = items.enumerated().flatMap { index, element in
                         return element.id == id ? index : nil
                     }
-                    indexesToRemove.dropFirst().reverse().forEach(items.removeAtIndex)
+                    indexesToRemove.dropFirst().reversed().forEach { index in items.remove(at: index) }
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks Shared/SharedFunctionality.swift]: `forEach(items.remove(at:))`?

File: RealmTasks Apple/RealmTasks Shared/SharedFunctionality.swift [hunk 11]
```diff
@@ -70,29 +70,29 @@ private func setDefaultRealmConfigurationWithUser(user: SyncUser) {
 
 // returns true on success
 func configureDefaultRealm() -> Bool {
-    if let user = SyncUser.currentUser() {
-        setDefaultRealmConfigurationWithUser(user)
+    if let user = SyncUser.current {
+        setDefaultRealmConfigurationWithUser(user: user)
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks Shared/SharedFunctionality.swift]: IMO this function should be renamed so the call site is `setDefaultRealmConfiguration(user: user)`

File: RealmTasks Apple/RealmTasks Shared/SharedFunctionality.swift [hunk 12]
```diff
@@ -70,29 +70,29 @@ private func setDefaultRealmConfigurationWithUser(user: SyncUser) {
...
     return false
 }
 
-func authenticate(username username: String, password: String, register: Bool, callback: (NSError?) -> ()) {
-    SyncUser.logInWithCredentials(.usernamePassword(username, password: password, register: register),
-                                  authServerURL: Constants.syncAuthURL) { user, error in
-        dispatch_async(dispatch_get_main_queue()) {
+
+func authenticate(username: String, password: String, register: Bool, callback: @escaping (NSError?) -> ()) {
+    SyncUser.logIn(with: SyncCredentials.usernamePassword(username: username, password: password, register: register), server: Constants.syncAuthURL) { user, error in
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks Shared/SharedFunctionality.swift]: `SyncUser.logIn(with: .usernamePassword(...`

File: RealmTasks Apple/RealmTasks iOS/CellPresenter.swift [hunk 13]
```diff
@@ -32,35 +32,37 @@ class CellPresenter<Item: Object where Item: CellPresentable> {
     }
 
     func deleteItem(item: Item) {
-        viewController.uiWrite {
-            guard !(item as Object).invalidated, let index = items.indexOf(item) else {
+        try! items.realm?.write {
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks iOS/CellPresenter.swift]: the whole idea behind `uiWrite` is to consolidate common things we want to do when doing interface-driven writes, such as the call to `didUpdateList` you've added a little further. Why did this change from `viewController.uiWrite` to `try! items.realm?.write`?

File: RealmTasks Apple/RealmTasks iOS/CellPresenter.swift [hunk 14]
```diff
@@ -32,35 +32,37 @@ class CellPresenter<Item: Object where Item: CellPresentable> {
...
-            viewController.tableView.deleteRowsAtIndexPaths([NSIndexPath(forRow: index, inSection: 0)], withRowAnimation: .Left)
+            viewController.tableView.deleteRows(at: [IndexPath(row: index, section: 0)], with: .left)
+            viewController.didUpdateList(reload: true)
         }
     }
 
     func completeItem(item: Item) {
-        viewController.uiWrite {
-            guard !(item as Object).invalidated, let index = items.indexOf(item) else {
+        try! items.realm?.write {
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks iOS/CellPresenter.swift]: `uiWrite`?

File: RealmTasks Apple/RealmTasks iOS/CellPresenter.swift [hunk 15]
```diff
@@ -32,35 +32,37 @@ class CellPresenter<Item: Object where Item: CellPresentable> {
...
                 // move cell just above the first completed item
                 let completedCount = items.filter("completed = true").count
-                destinationIndexPath = NSIndexPath(forRow: items.count - completedCount - 1, inSection: 0)
+                destinationIndexPath = IndexPath(row: items.count - completedCount - 1, section: 0)
             }
 
             items.move(from: sourceIndexPath.row, to: destinationIndexPath.row)
-            viewController.tableView.moveRowAtIndexPath(sourceIndexPath, toIndexPath: destinationIndexPath)
+            viewController.tableView.moveRow(at: sourceIndexPath, to: destinationIndexPath)
+            viewController.didUpdateList(reload: false)
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks iOS/CellPresenter.swift]: `uiWrite`?

File: RealmTasks Apple/RealmTasks iOS/CellPresenter.swift [hunk 16]
```diff
@@ -70,26 +72,26 @@ class CellPresenter<Item: Object where Item: CellPresentable> {
...
-        currentlyEditingIndexPath = tableView.indexPathForCell(editingCell)
+        currentlyEditingIndexPath = tableView.indexPath(for: editingCell)
 
-        let editingOffset = editingCell.convertRect(editingCell.bounds, toView: tableView).origin.y - tableView.contentOffset.y - tableView.contentInset.top
+        let editingOffset = editingCell.convert(editingCell.bounds, to: tableView).origin.y - tableView.contentOffset.y - tableView.contentInset.top
         viewController.setTopConstraintTo(constant: -editingOffset)
         tableView.contentInset.bottom += editingOffset
 
-        viewController.setPlaceholderAlpha(0)
+        viewController.setPlaceholderAlpha(alpha: 0)
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks iOS/CellPresenter.swift]: method should be refactored to be `setPlaceholderAlpha(0)` at the call site IMO.

File: RealmTasks Apple/RealmTasks iOS/CellPresenter.swift [hunk 17]
```diff
@@ -106,26 +108,26 @@ class CellPresenter<Item: Object where Item: CellPresentable> {
...
             }
             tableView.superview?.layoutSubviews()
         }
 
         let item = editingCell.item
-        if item.text.isEmpty {
-            let indexPath = NSIndexPath(forRow: items.indexOf(item)!, inSection: 0)
-            item.realm!.delete(item)
-            viewController.tableView.deleteRowsAtIndexPaths([indexPath], withRowAnimation: .Top)
+        if (item?.text.isEmpty)! {
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks iOS/CellPresenter.swift]: this is a really funny way to unwrap an optional. `item!.text.isEmpty`. Although I think the problem should actually be solved at the source, which is that `editingCell.item` shouldn't be optional.

File: RealmTasks Apple/RealmTasks iOS/ListPresenter.swift [hunk 18]
```diff
@@ -61,36 +61,36 @@ class ListPresenter<Item: Object, Parent: Object where Item: CellPresentable, Pa
 
     func observeListTitle() {
         if let parent = parent as? CellPresentable {
-            (parent as! Object).addObserver(self, forKeyPath: "text", options: .New, context: &titleKVOContext)
-            viewController.setListTitle(parent.text)
+            (parent as! Object).addObserver(self, forKeyPath: "text", options: .new, context: &titleKVOContext)
+            viewController.setListTitle(title: parent.text)
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks iOS/ListPresenter.swift]: should be refactored so that the call site is: `setListTitle(parent.text)`.

File: RealmTasks Apple/RealmTasks iOS/LogInViewController/Classes/UI/LogInViewController.swift [hunk 19]
```diff
@@ -26,15 +26,15 @@ enum LogInViewControllerReturnCode: Int {
 
 class LogInViewController: UIViewController {
 
-    @IBOutlet private weak var userNameTextField: UITextField!
-    @IBOutlet private weak var passwordTextField: UITextField!
-    @IBOutlet private weak var logInButton: UIButton!
+    @IBOutlet weak var userNameTextField: UITextField!
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks iOS/LogInViewController/Classes/UI/LogInViewController.swift]: should these all be `private(set)`?

File: RealmTasks Apple/RealmTasks iOS/LogInViewController/Classes/UI/RegisterViewController.swift [hunk 20]
```diff
@@ -25,20 +25,20 @@ enum RegisterViewControllerReturnCode: Int {
 
 class RegisterViewController: UIViewController {
 
-    @IBOutlet private weak var userNameTextField: UITextField!
-    @IBOutlet private weak var passwordTextField: UITextField!
-    @IBOutlet private weak var confirmationTextField: UITextField!
-    @IBOutlet private weak var registerButton: UIButton!
+    @IBOutlet weak var userNameTextField: UITextField!
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks iOS/LogInViewController/Classes/UI/RegisterViewController.swift]: should these all be `private(set)`?

File: RealmTasks Apple/RealmTasks iOS/TablePresenter.swift [hunk 21]
```diff
@@ -248,15 +248,15 @@ class TablePresenter<Parent: Object where Parent: ListPresentable>: NSObject,
...
-        let location = gestureRecognizer.locationInView(tableView)
-        if let indexPath = tableView.indexPathForRowAtPoint(location),
-            cell = tableView.cellForRowAtIndexPath(indexPath) as? TableViewCell<Parent.Item> {
+        let location = gestureRecognizer.location(in: tableView)
+        if let indexPath = tableView.indexPathForRow(at: location),
+            let cell = tableView.cellForRow(at: indexPath) as? TableViewCell<Parent.Item> {
             return !cell.item.completed
         }
-        return gestureRecognizer.isKindOfClass(UITapGestureRecognizer.self)
+        return (gestureRecognizer is UITapGestureRecognizer)
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks iOS/TablePresenter.swift]: superfluous parens

File: RealmTasks Apple/RealmTasks iOS/TablePresenter.swift [hunk 22]
```diff
@@ -265,17 +265,17 @@ class TablePresenter<Parent: Object where Parent: ListPresentable>: NSObject,
 
     func colorForRow(row: Int) -> UIColor {
         let fraction = Double(row) / Double(max(13, items.count))
-        return colors.gradientColorAtFraction(fraction)
+        return colors.gradientColorAtFraction(fraction: fraction)
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks iOS/TablePresenter.swift]: `return colors.gradientColor(atFraction: fraction)`

File: RealmTasks Apple/RealmTasks iOS/TablePresenter.swift [hunk 23]
```diff
@@ -287,11 +287,11 @@ class TablePresenter<Parent: Object where Parent: ListPresentable>: NSObject,
     // MARK: Placeholder cell
 
     // Placeholder cell to use before being adding to the table view
-    private let placeHolderCell = TableViewCell<Parent.Item>(style: .Default, reuseIdentifier: "cell")
+    private let placeHolderCell = TableViewCell<Parent.Item>(style: .default, reuseIdentifier: "cell")
 
     func setupPlaceholderCell(inTableView tableView: UITableView) {
         placeHolderCell.alpha = 0
-        placeHolderCell.backgroundView!.backgroundColor = colorForRow(0)
+        placeHolderCell.backgroundView!.backgroundColor = colorForRow(row: 0)
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks iOS/TablePresenter.swift]: `= color(forRow: 0)`

File: RealmTasks Apple/RealmTasks iOS/TablePresenter.swift [hunk 24]
```diff
@@ -318,25 +318,25 @@ class TablePresenter<Parent: Object where Parent: ListPresentable>: NSObject,
                 let angle = CGFloat(M_PI_2) - tan(distancePulledDown / cellHeight)
 
                 var transform = CATransform3DIdentity
-                transform.m34 = 1 / -(1000 * 0.2)
+                transform.m34 = CGFloat(1.0 / -(1000 * 0.2))
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks iOS/TablePresenter.swift]: these changes shouldn't be necessary

File: RealmTasks Apple/RealmTasks iOS/TablePresenter.swift [hunk 25]
```diff
@@ -318,25 +318,25 @@ class TablePresenter<Parent: Object where Parent: ListPresentable>: NSObject,
...
-                placeHolderCell.navHintView.hintArrowTransfom = CGAffineTransformRotate(CGAffineTransformIdentity, CGFloat(M_PI))
+                placeHolderCell.navHintView.hintArrowTransfom = CGAffineTransform.identity.rotated(by: CGFloat(M_PI))
 
-                UIView.animateWithDuration(0.4, delay: 0, usingSpringWithDamping: 1, initialSpringVelocity: 0.5,
+                UIView.animate(withDuration: 0.4, delay: 0, usingSpringWithDamping: 1, initialSpringVelocity: 0.5,
                     options: [], animations: { [unowned self] in
 
                     self.placeHolderCell.navHintView.alpha = 1
-                    self.placeHolderCell.navHintView.hintArrowTransfom  = CGAffineTransformIdentity
+                    self.placeHolderCell.navHintView.hintArrowTransfom  = CGAffineTransform.identity
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks iOS/TablePresenter.swift]: `hintArrowTransfom  = .identity`?

File: RealmTasks Apple/RealmTasks iOS/TableViewCell.swift [hunk 26]
```diff
@@ -54,7 +54,7 @@ final class TableViewCell<Item: Object where Item: CellPresentable>: UITableView
     var item: Item! {
         didSet {
             textView.text = item.text
-            setCompleted(item.completed)
+            setCompleted(completed: item.completed)
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks iOS/TableViewCell.swift]: `setCompleted(item.completed)`

File: RealmTasks Apple/RealmTasks iOS/TableViewCell.swift [hunk 27]
```diff
@@ -319,7 +319,7 @@ final class TableViewCell<Item: Object where Item: CellPresentable>: UITableView
                 self.deleteIconView.frame.origin.x = -iconWidth + self.deleteIconView.bounds.width + 20
             }
             completionBlock = {
-                self.presenter.deleteItem(self.item)
+                self.presenter.deleteItem(item: self.item)
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks iOS/TableViewCell.swift]: `self.presenter.delete(item: self.item)`

File: RealmTasks Apple/RealmTasks iOS/ViewController.swift [hunk 28]
```diff
@@ -69,12 +69,12 @@ final class ViewController<Item: Object, Parent: Object where Item: CellPresenta
 
     // MARK: UI Writes
 
-    func uiWrite(@noescape block: () -> ()) {
-        uiWriteNoUpdateList(block)
+    func uiWrite( block: () -> ()) {
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks iOS/ViewController.swift]: spacing

File: RealmTasks Apple/RealmTasks iOS/ViewController.swift [hunk 29]
```diff
@@ -69,12 +69,12 @@ final class ViewController<Item: Object, Parent: Object where Item: CellPresenta
...
+        uiWriteNoUpdateList(block: block)
         didUpdateList(reload: false)
     }
 
-    func uiWriteNoUpdateList(@noescape block: () -> ()) {
+    func uiWriteNoUpdateList( block: () -> ()) {
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks iOS/ViewController.swift]: spacing

File: RealmTasks Apple/RealmTasks iOS/ViewController.swift [hunk 30]
```diff
@@ -228,60 +232,60 @@ final class ViewController<Item: Object, Parent: Object where Item: CellPresenta
...
+            if bottomViewController === parent?.childViewControllers.last { return }
             if bottomViewController == nil {
                 bottomViewController = createAuxController()
             }
-            startMovingToNextViewController(.Down)
+            startMovingToNextViewController(direction: .Down)
             return
         } else {
-            removeVC(bottomViewController)
+            removeVC(viewController: bottomViewController)
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks iOS/ViewController.swift]: either remove the `VC` or `viewController` label. This naming is redundant.

File: RealmTasks Apple/RealmTasks Shared/RealmColor.swift (hunk 9 above)
Comment (TimOliver) [inline RealmTasks Apple/RealmTasks Shared/RealmColor.swift]: It was simply `Double(self.count)` in Swift 2, but since `self.count` was changed to type `IndexDistance` in Swift 3, that wouldn't compile.

I looked into it further and it seems `self.endIndex` is an appropriate replacement now.

File: RealmTasks Apple/RealmTasks Shared/SharedFunctionality.swift (hunk 10 above)
Comment (TimOliver) [inline RealmTasks Apple/RealmTasks Shared/SharedFunctionality.swift]: How would that look exactly? In any case, I had a crack at simplifying it.

File: RealmTasks Apple/RealmTasks iOS/LogInViewController/Classes/UI/LogInViewController.swift (hunk 19 above)
Comment (TimOliver) [inline RealmTasks Apple/RealmTasks iOS/LogInViewController/Classes/UI/LogInViewController.swift]: Ah, no. They apparently needed to be set as `fileprivate`.

File: RealmTasks Apple/RealmTasks iOS/TablePresenter.swift (hunk 24 above)
Comment (TimOliver) [inline RealmTasks Apple/RealmTasks iOS/TablePresenter.swift]: Agreed, but this was the resulting build error I was getting without explicitly casting to `CGFloat`:

screen shot 2017-01-12 at 12 37 20 am

File: RealmTasks Apple/RealmTasks iOS/CellPresenter.swift (hunk 13 above)
Comment (TimOliver) [inline RealmTasks Apple/RealmTasks iOS/CellPresenter.swift]: Oh dear. I think this was a merge error when I updated the branch. I've updated them to `uiWrite` now.

File: RealmTasks Apple/RealmTasks iOS/CellPresenter.swift (hunk 15 above)
Comment (TimOliver) [inline RealmTasks Apple/RealmTasks iOS/CellPresenter.swift]: This block of code is already inside a `uiWrite` block. Is it necessary to encapsulate it again?

**Comment by TimOliver** [comment]
Thanks for the input @stel and @jpsim! I've followed up the PR with all of your feedback. Let me know what you think!

File: RealmTasks Apple/RealmTasks iOS/CellPresenter.swift [hunk 31]
```diff
@@ -31,36 +31,38 @@ class CellPresenter<Item: Object where Item: CellPresentable> {
...
         viewController.uiWrite {
-            guard !(item as Object).invalidated, let index = items.indexOf(item) else {
+            guard !(item as Object).isInvalidated, let index = items.index(of: item) else {
                 return
             }
 
             items.realm?.delete(item)
-            viewController.tableView.deleteRowsAtIndexPaths([NSIndexPath(forRow: index, inSection: 0)], withRowAnimation: .Left)
+            viewController.tableView.deleteRows(at: [IndexPath(row: index, section: 0)], with: .left)
+            viewController.didUpdateList(reload: true)
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks iOS/CellPresenter.swift]: why was this call to `didUpdateList` added? It wasn't there before. Was that a previous omission, or was this added accidentally?

Comment (TimOliver) [inline RealmTasks Apple/RealmTasks iOS/CellPresenter.swift]: I just tested it out to see what it was doing; that API call is in charge of updating the cell background colors and general UI state, so it's definitely needed in there.

`(reload: true)` however isn't needed anymore so I changed it to `false` and tested animations are properly playing now.

**Comment by TimOliver** [comment]
Thanks a lot @stel! No I haven't got anything specific that needs checking anymore. :)

File: RealmTasks Apple/RealmTasks iOS/CellPresenter.swift [hunk 32]
```diff
@@ -31,36 +31,38 @@ class CellPresenter<Item: Object where Item: CellPresentable> {
...
         viewController.uiWrite {
-            guard !(item as Object).invalidated, let index = items.indexOf(item) else {
+            guard !item.isInvalidated, let index = items.index(of: item) else {
                 return
             }
 
             items.realm?.delete(item)
-            viewController.tableView.deleteRowsAtIndexPaths([NSIndexPath(forRow: index, inSection: 0)], withRowAnimation: .Left)
+            viewController.tableView.deleteRows(at: [IndexPath(row: index, section: 0)], with: .left)
+            viewController.didUpdateList(reload: false)
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks iOS/CellPresenter.swift]: I think my previous comment on this was lost. Why was this call to `didUpdateList(...)` added here? Was there a bug in the Swift 2.3 version of this code? Or was this added by mistake?

Comment (TimOliver) [inline RealmTasks Apple/RealmTasks iOS/CellPresenter.swift]: Okay. I went through the commit history with a fine-toothed comb to see what happened.

You originally removed `didUpdateList` when you added the `uiWrite` blocks to both the `delete` and `complete` methods since that call is now made internally at the end of that block. When I merged in the latest changes from the `master` branch, they got added back in.

Since the update calls are no longer necessary in either method, I removed them.

Comment (jpsim) [inline RealmTasks Apple/RealmTasks iOS/CellPresenter.swift]: thanks for the update!
